### ISSUE TITLE
False positive ``__init__ method from base class is not called`` when the constructor raises ``NotImplementedError`` without using ``abc``

### ISSUE THREAD

**Issue by cpitclaudel** [PyCQA/pylint#3975]
### Is your feature request related to a problem? Please describe

In the following example, Pylint flags the `Derived`'s `__init__` method because it doesn't call `Bases`'s `__init__`.

```python
class Base:
    def __init__(self, a: int, b: str):
        raise NotImplementedError()

class Derived(Base):
    def __init__(self, a: int, b: str):
        self.a = a + 1
        self.b = b[::-1]
```

This is discussed here: https://stackoverflow.com/questions/25072363/error-init-method-from-base-class-is-not-called-for-an-abstract-class

### Describe the solution you'd like

Pylint could (should?) detect this case and not report an error.

### Additional context

An easy fix is to remove the init in the base class, but it can be useful to have it there to define what signature `__init__` should have.

**Comment by hippo91** [comment]
@cpitclaudel thanks for your report. Why not using `abc` module and defining the base class as an abstract one?

**Comment by cpitclaudel** [comment]
Are you thinking of this?

```python
from abc import ABCMeta, abstractmethod

class Base:
    __metaclass__ = ABCMeta

    @abstractmethod
    def __init__(self, a: int, b: str):
        pass

class Derived(Base):
    def __init__(self, a: int, b: str):
        super().__init__(a, b)
        self.a = a + 1
        self.b = b[::-1]
```

That's what one of the answers on stackoverflow suggests, but it's a lot of cruft, and it doesn't prevent `Base` from being instantiated.

**Comment by cpitclaudel** [comment]
> Why not using abc module and defining the base class as an abstract one?

A different perspective: maybe there *should* be a pylint rule that suggests moving to an ABS instead of using NotImplementedError(); but this issue is about a different problem, namely pylint suggesting to call an __init__ that raises NotImplementedError/

**Comment by hippo91** [comment]
@cpitclaudel indeed if you use metaclass this way:
```
from abc import ABCMeta, abstractmethod

class Base(metaclass=ABCMeta):
    @abstractmethod
    def __init__(self, a: int, b: str):
        pass

class Derived(Base):
    def __init__(self, a: int, b: str):
        super().__init__(a, b)
        self.a = a + 1
        self.b = b[::-1]


if __name__ == "__main__":
    x = Base(3, "toto")
```
and execute this program, you will get:
```
TypeError: Can't instantiate abstract class Base with abstract methods __init__
```
Thus it is impossible to instantiate `Base` class.

However i agree that `pylint` should be able to detect that the `__init__` method of the base class raises `NotImplementError` and thus should invite the user to switch toward use of `ABCMeta`.

**Comment by cpitclaudel** [comment]
> However i agree that pylint should be able to detect that the __init__ method of the base class raises NotImplementError and thus should invite the user to switch toward use of ABCMeta.

:+1: , though does the added ABCMeta stuff buy anything compared to the version that just throws an exception?  Pylint currently recognizes methods that throw `NotImplementedError`s just fine, without recommending to move these too.

And having to call __init__ in the derived class is definitely a downside, no?

**Comment by Py-GNU-Unix** [comment]
I agree with the need to ignore `super-init-not-called` if the super init raises `NotImplementedError`. That's a important feature, but no one fixed this in a long time. I tried to fix this by myself, but that's the first time that i look into pylint code and i failed. I hope that someone will fix this...

**Comment by Py-GNU-Unix** [comment]
might someone add an high priority tag to this issue? This is nearly a bug.

**Comment by Pierre-Sassoulas** [comment]
Hi @Py-GNU-Unix The high priority tag is for issue that have more than 10 thumbsup/heart reactions, but we're reviewing every contribution. If you want to contribute to this issue, it's going to get merged before high priority issues that do not get any attention.

**Comment by vivi90** [comment]
I think it's **NOT** an **False Positive**.
The reason, i think so is:
 * __If you are NOT USING the `abc` package:__ Might be hard to detect abstract methods in a reliable way. So `pylint` prevents you here from possible logical issues.
 * __If you are USING the `abc` package:__ Even if you are not able to instantiate the base class without implementing all abstract methods (including `__init__`) in an concrete way, you ARE INDEED ABLE to implement some stuff in abstract methods of the base class and call this methods with `super()` in it's subclass(es). For the `__init__` method this means, that `pylint` prevents you here from violating OOP concepts.
 
An *good practice* example from me:
```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""Abstract class example."""

from abc import ABCMeta, abstractmethod


class Abstract(metaclass=ABCMeta):
    """Abstract class."""

    @property
    @abstractmethod
    def message(self) -> str:
        """str: Return message."""

    @abstractmethod
    def __init__(self, message: str) -> None:
        """Abstract constructor.

        Args:
            message (str):
                Message.

        """


class Concrete(Abstract):
    """Concrete class."""

    @property
    def message(self) -> str:
        """str: Return message."""
        return self._message

    def __init__(self, message: str) -> None:
        """Concrete constructor.

        Args:
            message (str):
                Message.

        """
        super().__init__(message)
        self._message = message


print(Concrete("test").message)

```
*(using `class Abstract(metaclass=ABCMeta)` instead of `class Abstract(ABC)` here, because it's more explicit and works also well in combination with `Protocol` [if really needed])*

**Comment by cpitclaudel** [comment]
@vivi90 Your example is great, but this issue is about an `__init__` that throws a `NotImplementedError`.  `pylint` suggests adding a call to that `__init__`, which would throws a runtime exception.  That's not good ^^

**Comment by vivi90** [comment]
@cpitclaudel
> [..] `pylint` suggests adding a call to that `__init__`, which would throws a runtime exception. [..]

Yes, thats **WRONG behaviour** but with the **RIGHT goal**. 😄
So how about adding an new warning message like `use 'abc' instead of raising 'NotImplementedError' here` in combination with the still existing `'__init__' method from base class is not called` message instead of silencing `pylint` in this case? 🤔

**Comment by DanielNoord** [comment]
> @cpitclaudel
> 
> > [..] `pylint` suggests adding a call to that `__init__`, which would throws a runtime exception. [..]
> 
> 
> 
> Yes, thats **WRONG behaviour** but with the **RIGHT goal**. 😄
> 
> So how about adding an new warning message like `use 'abc' instead of raising 'NotImplementedError' here` in combination with the still existing `'__init__' method from base class is not called` message instead of silencing `pylint` in this case? 🤔 
> 
> 

I think it makes sense to remove abstract methods from this check. Due to historic reasons `pylint` considers a number of patterns to be "abstract". For example, the default setting is to consider `pass` abstract as well.
We probably shouldn't change this.

That said, I think an argument can be made to create a code style check that suggests to inherit from `abc`. That should probably be its own issue though.

**Comment by vivi90** [comment]
> I think it makes sense to remove abstract methods from this check. [..]

I think that everything might be better than removing this check for abstract `__init__` methods, because of the second point of [my explanation above](https://github.com/PyCQA/pylint/issues/3975#issuecomment-1193362030).

> [..] We probably shouldn't change this.

Yes, i think the same.

> [..] an argument can be made to create a code style check that suggests to inherit from `abc`. [..]

Yes, exactly.

> [..] That should probably be its own issue though.

Discuissing this in an own issue might be out of context.
Here we have an good case at hand.

**Comment by DanielNoord** [comment]
> I think that everything might be better than removing this check for abstract `__init__` methods, because of the second point of [my explanation above](https://github.com/PyCQA/pylint/issues/3975#issuecomment-1193362030).

Whenever you implement anything in the abstract class but also raise UnimplementedError I don't think pylint should be able to recognise that. That's just an anti-pattern.

Pylint's recognition of abstract methods and classes is actually quite old and robust so I don't think that should limit us.

We tend to separate issues and messages from each other: users might not have all messages turned on so dependencies between messages are best avoided.

Because of this we should take this message and ask ourselves: should `super-init-not-called` be raised on abstract `__init__`? I think the answer is no as if there is anything in the `__init__` it is not actually abstract.

If this then creates a false positive or negative we should fix our abstract recognition, but the premise of not raising is still valid imo.

**Comment by vivi90** [comment]
@DanielNoord 
> Whenever you implement anything in the abstract class but also raise UnimplementedError I don't think pylint should be able to recognise that. That's just an anti-pattern.

Oh no, that's not what i wanted to purpose. 😆

> We tend to separate issues and messages from each other: users might not have all messages turned on so dependencies between messages are best avoided.

Yes, that's also not what i wanted to say.
The `'__init__' method from base class is not called` should not be an dependecy for the other message or the other way around.

> Because of this we should take this message and ask ourselves: should `super-init-not-called` be raised on abstract `__init__`? I think the answer is no as if there is anything in the `__init__` it is not actually abstract.

hm okay so you mean if someone write something like this..
```python
    @abstractmethod
    def __init__(self, message: str) -> None:
        """Abstract constructor.

        Args:
            message (str):
                Message.

        """
        this._message = message
```
then this `__init__` method is not abstract anymore? 🤔 
I mean, it's possible to do that and using it in the subclass by `super()`.
That's what i pointed to with:
> I think that everything might be better than removing this check for abstract __init__ methods, because of the second point of [my explanation above](https://github.com/PyCQA/pylint/issues/3975#issuecomment-1193362030).

> If this then creates a false positive or negative we should fix our abstract recognition, but the premise of not raising is still valid imo.

Hmm, for `abc` it seems still abstract, since there is an `@abstractmethod` decorator. 🤔

**Comment by DanielNoord** [comment]
> hm okay so you mean if someone write something like this..
> 
> ```python
> 
>     @abstractmethod
> 
>     def __init__(self, message: str) -> None:
> 
>         """Abstract constructor.
> 
> 
> 
>         Args:
> 
>             message (str):
> 
>                 Message.
> 
> 
> 
>         """
> 
>         this._message = message
> 
> ```
> 
> then this `__init__` method is not abstract anymore? 🤔 

No, as long as something "meaningful" happens in the method it is not considered abstract.

> Hmm, for `abc` it seems still abstract, since there is an `@abstractmethod` decorator. 🤔 

Yeah. So in such cases this shouldn't warn.

**Comment by vivi90** [comment]
@DanielNoord 
> Yeah. So in such cases this shouldn't warn.

That's exactly the point i am worried about.
If it does NOT warn in this case, it may happen that:
 1. The user overwrites this `__init__` method in the subclass, because it's abstract according to `abc`.
 2. The user doesn't call `super().__init__(...)` in the overwritten `__init__` method.
 3. The logic in the 'abstract' `__init__` method of the base class is never called..

So then i prefer the current behaviour of `pylint`. Please don't change it.

**Comment by DanielNoord** [comment]
I have added a test to #7227 to make the new behaviour more explicit. It should be:

```python
class Base:
    def __init__(self, param: int, param_two: str) -> None:
        raise NotImplementedError()


class Derived(Base):
    def __init__(self, param: int, param_two: str) -> None:
        self.param = param + 1
        self.param_two = param_two[::-1]


class AbstractBase(abc.ABC):
    def __init__(self, param: int) -> None:
        self.param = param + 1

    def abstract_method(self) -> str:
        """This needs to be implemented."""
        raise NotImplementedError()


class DerivedFromAbstract(AbstractBase):
    def __init__(self, param: int) -> None:  # [super-init-not-called]
        print("Called")

    def abstract_method(self) -> str:
        return "Implemented"
```

Whenever ``__init__`` is abstract itself we don't raise when it isn't called. If the class is abstract but the `__init__` isn't, we do raise.

**Comment by cpitclaudel** [comment]
Thanks for your work @DanielNoord

**Comment by DanielNoord** [comment]
#7227 was merged so this should be available in `2.15` for all those watching this issue 😄
